### PR TITLE
libunwind: Add 1.7.2, 1.8.1, and new *-stable branches

### PR DESCRIPTION
**Pull request by msimberg**
Also changes the download URL to use the github repo: https://github.com/libunwind/libunwind. Releases after 1.6.X are no longer available on https://download.savannah.nongnu.org/releases/libunwind/.

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
Confirmed the branches and all but two sha256s.  I am unable to confirm `1.5.0` and `1.1`.  You can add the old URL for those two versions as arguments *or* update `url_for_version` to not prepend `v` to the version numbers for those two versions.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
See prior comment.

**Comment by tldahlgren** [comment]
Looks like a couple of CI stacks are failing, at least one of which on `libunwind`.

**Comment by mwkrentel** [comment]
A few things:

1. I think the github site has tarballs as far back as 1.2.1.  I think
it makes sense to switch over to github for things earlier than 1.7 as
well as the new ones.

2. In view of (1), version `1.1` is already deprecated, we could just
drop it if that makes the URLs simpler.

3. The unfortunate outlier is 1.5.0.  To fit the pattern on github, the
URL should be:
```
https://github.com/libunwind/libunwind/releases/download/v1.5.0/libunwind-1.5.0.tar.gz
```
But it's actually this, with the first `v1.5.0` shortened to `v1.5`
```
https://github.com/libunwind/libunwind/releases/download/v1.5/libunwind-1.5.0.tar.gz
```
Grrr, how annoying.  I guess this is reason enough to keep `url_for_version`
for one-off exceptions like this.  Blech.  :-)

4. Could you add branches for `1.7-stable` and `1.8-stable`.  I think
they fit the same pattern as `1.6` and `1.5`.
We could drop the `1.5-stable` branch.   That was my plan all along
to keep the two most recent stable branches plus master.
But since you're adding two releases, we'll also keep 1.6-stable for now.

5. The failures of `1.8.1` on powerpc and aarch astound me.  There are
issues on github about aarch build failures, but not for powerpc.
Maybe we have to avoid adding `1.8.1` until those are resolved, or at
least add `1.8.1` only for x86.

@msimberg Thanks for doing this.  I've been checking the savannah download site,
but I missed that they moved on to github.

**Comment by msimberg** [comment]
Thank you @tldahlgren and @mwkrentel for reviewing!

>     1. I think the github site has tarballs as far back as 1.2.1.  I think
>        it makes sense to switch over to github for things earlier than 1.7 as
>        well as the new ones.

Good idea. I changed it to use github for all releases. Checksums were even the same for the older releases so they stay unchanged.

>     2. In view of (1), version `1.1` is already deprecated, we could just
>        drop it if that makes the URLs simpler.

Removed.
 
>     3. The unfortunate outlier is 1.5.0.  To fit the pattern on github, the
>        URL should be:
>
>        Blech. :-)

Blech indeed :smile: I've added the special case for 1.5.0...
 
>     4. Could you add branches for `1.7-stable` and `1.8-stable`.  I think
>        they fit the same pattern as `1.6` and `1.5`.
>        We could drop the `1.5-stable` branch.   That was my plan all along
>        to keep the two most recent stable branches plus master.
>        But since you're adding two releases, we'll also keep 1.6-stable for now.

Added 1.7 and 1.8, removed 1.5. However... The build actually fails for 1.7-stable, 1.8-stable, and master, all with the same error:

```
==> libunwind: Executing phase: 'autoreconf'
==> Error: ProcessError: Command exited with status 1:
    '/home/mjs/src/spack/opt/spack/linux-nixos24-icelake/gcc-14.1.0/autoconf-2.72-aoyafxdm2wv6mzvk3jnx3dxwqjqfo7hg/bin/autoreconf' '-ivf' '-I' '/home/mjs/src/spack/opt/spack/linux-nixos24-icelake/gcc-14.1.0/libtool-2.4.7-axtq6en62ebg2xhjljdcu3jdwkbx4ryu/share/aclocal' '-I' '/home/mjs/.nix-profile/share/aclocal'

2 errors found in build log:
     6     ==> [2024-11-05-14:09:05.827580] Warning: *********************************************************
     7     ==> [2024-11-05-14:09:05.828028] '/home/mjs/src/spack/opt/spack/linux-nixos24-icelake/gcc-14.1.0/autoconf-2.72-aoya
           fxdm2wv6mzvk3jnx3dxwqjqfo7hg/bin/autoreconf' '-ivf' '-I' '/home/mjs/src/spack/opt/spack/linux-nixos24-icelake/gcc-1
           4.1.0/libtool-2.4.7-axtq6en62ebg2xhjljdcu3jdwkbx4ryu/share/aclocal' '-I' '/home/mjs/.nix-profile/share/aclocal'
     8     autoreconf: export WARNINGS=
     9     autoreconf: Entering directory '.'
     10    autoreconf: configure.ac: not using Gettext
     11    autoreconf: running: aclocal -I /home/mjs/src/spack/opt/spack/linux-nixos24-icelake/gcc-14.1.0/libtool-2.4.7-axtq6e
           n62ebg2xhjljdcu3jdwkbx4ryu/share/aclocal -I /home/mjs/.nix-profile/share/aclocal --force -I m4
  >> 12    aclocal: error: couldn't open directory 'm4': No such file or directory
  >> 13    autoreconf: error: aclocal failed with exit status: 1
```

Does that look familiar to you? I've obviously got more detailed logs if that helps. I can try to look into this as well eventually but probably not this week.

>     5. The failures of `1.8.1` on powerpc and aarch astound me.  There are
>        issues on github about aarch build failures, but not for powerpc.
>        Maybe we have to avoid adding `1.8.1` until those are resolved, or at
>        least add `1.8.1` only for x86.

Yeah, strange... For PPC I think I found both the PR that introduced the breakage and the PR that fixed it (added as comments next to the conflict). I think this applies only to 1.8.0 and 1.8.1.

For aarch64 I couldn't find any specific commit that might have introduced the issue or fixed it, so I've attempted to add a `when="@1.8:"` conflict for aarch64 for the moment. If that's not enough we can bump it down to 1.7 (I'm assuming 1.6 is fine since that already was there).

For both of the above I think if there's interest they can be patched up later to build with any version, but I'd avoid doing that here.

**Comment by mwkrentel** [comment]
> Checksums were even the same for the older releases so they stay unchanged.

@msimberg I forgot about that.  I'm kinda surprised because I thought
github was creating the tar file on the fly.  Should double check that.

> The build actually fails for 1.7-stable, 1.8-stable, and master,

Oops, sorry, forgot to mention that you also need to update `reconf_versions`
in the following, but it's just one line.
```
# The libunwind releases contain the autotools generated files,
# but the git repo snapshots do not.
reconf_versions = "@master,1.5-stable,1.6-stable"
depends_on("autoconf", type="build", when=reconf_versions)
depends_on("automake", type="build", when=reconf_versions)
depends_on("libtool", type="build", when=reconf_versions)
depends_on("m4", type="build", when=reconf_versions)
```

`master` still fails?  I also fails for me.
```
     11    autoreconf: running: aclocal -I /scratch/krentel/spack/install/linux-rhel8-zen2/gcc-14.1.0/libtool-2.4.7-jmqap
           ztbggsvazc7wxofm74jehj35fkf/share/aclocal --force -I m4
  >> 12    aclocal: error: couldn't open directory 'm4': No such file or directory
  >> 13    autoreconf: error: aclocal failed with exit status: 1
```

It looks like `autoreconf` is expecting the `m4` directory to already exist,
but `libunwind` expects autoreconf to create it.
@tldahlgren is this a spack thing?  Do we have the autotools dependencies wrong?

As for the `powerpc` and `aarch64` failures, long term, we'd either
patch 1.8.1 or else wait for a release where it is fixed.
I'd have to check the status at libunwind.

Btw, I'd be tempted not to even generate the versions.  You can do
this, but it's a little bit "outside" spack, since you can't write
`when` inside `version`.
```
import platform
...
machine = platform.machine().lower()
if machine == "x86_64":
    version("1.8.1", sha256="...")
```
But maybe that's just me. :-)

**Comment by msimberg** [comment]
> @msimberg I forgot about that. I'm kinda surprised because I thought github was creating the tar file on the fly. Should double check that.

I suppose I chose the URL for archives that the maintainers have added to the release manually instead of the auto-generated archive, which is probably why they're unchanged. I can also change it to the auto-generated ones if you think that's better?

> > The build actually fails for 1.7-stable, 1.8-stable, and master,
> 
> Oops, sorry, forgot to mention that you also need to update `reconf_versions` in the following, but it's just one line.

I think It's not that, I happened to notice `reconf_versions` while I was changing things as well. That was up to date when I attempted to build those versions...

> As for the `powerpc` and `aarch64` failures, long term, we'd either patch 1.8.1 or else wait for a release where it is fixed. I'd have to check the status at libunwind.

:+1:

Yeah, patching them would be fine by me eventually. If you prefer, I could backport the PPC patch already now? For aarch64 there's nothing to backport though...
 
> Btw, I'd be tempted not to even generate the versions. You can do this, but it's a little bit "outside" spack, since you can't write `when` inside `version`.
> 
> ```
> import platform
> ...
> machine = platform.machine().lower()
> if machine == "x86_64":
>     version("1.8.1", sha256="...")
> ```
> 
> But maybe that's just me. :-)

I'm mostly indifferent, but obviously I slightly prefer the `conflicts` (a `version` with `when` would be even nicer, but as you said that's not possible). I think doing an `if machine == ...` also doesn't do the right thing for cross-compilation?

**Comment by msimberg** [comment]
Looks like the pipelines succeeded, so the current conflicts are at least sufficient. Let me know if you'd like to see the conflicts implemented differently or as patches (at least for the PPC conflict) instead.

**Review by mwkrentel — APPROVED** [review]
I'm Ok with these additions and we'll leave fixing the master/stable
branches and 1.8 for powerpc and aarch64 for another day.
I approve.
@tldahlgren  ?

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**
